### PR TITLE
FOAM FObject ⇔ CSV Parsing

### PR DESCRIPTION
**Pull request by TanayParikh**
- Implements two directional FObject to CSV parsing.

### Can be used with the following commands
```js
foam.lib.csv.stringify(FObject)
```

```js
foam.lib.csv.objectify(String)
```

## ⇔
```js
foam.lib.csv.objectify(foam.lib.csv.stringify(FObject))
```

### CODE REVIEW

File: src/foam/core/CSV.js [hunk 1]
```diff
@@ -0,0 +1,382 @@
...
+          for ( var i = 0 ; i < ps.length ; i++ ) {
+            this.outputProperty(o, ps[i]);
+          }
+
+          this.removeTrailingComma();
+        },
+        Array: function(o) { /* Ignore arrays in CSV */ },
+        Function: function(o) { /* Ignore functions in CSV */ },
+        Object: function(o) {
+          debugger;
```
Comment (kirkeaton) [inline src/foam/core/CSV.js]: Remove debugger statement

Comment (TanayParikh) [reply src/foam/core/CSV.js]: Thanks, removed and replaced with TODO

File: src/foam/core/CSV.js [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+  properties: [
+    {
+      name: 'fromCSV',
+      value: function fromCSV(value, ctx, prop, csv) {
+        return foam.csv.parse(value, null, ctx);
+      }
+    },
+    {
+      name: 'toCSV',
+      value: function toCSV(value, outputter) { return value; }
```
Comment (kgrgreer) [inline src/foam/core/CSV.js]: What is the outputter used for?

File: src/foam/core/CSV.js [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+      value: '\n'
+    },
+    {
+      class: 'String',
+      name: 'undefinedStr',
+      value: ''
+    },
+    {
+      class: 'Boolean',
+      name: 'outputDefaultValues',
```
Comment (kgrgreer) [inline src/foam/core/CSV.js]: I don't think this works for CSV, since the location is always present, making it impossible to tell the difference between the default value and an empty string.

File: src/foam/core/CSV.js [hunk 4]
```diff
@@ -0,0 +1,382 @@
...
+    // TODO: Copied from foam.core.json (move to common library)
+    function nl() {
+      if ( this.nlStr && this.nlStr.length ) {
+        this.out(this.nlStr);
+      }
+
+      return this;
+    },
+
+    // TODO: Copied from foam.core.json (move to common library)
```
Comment (kgrgreer) [inline src/foam/core/CSV.js]: Do these still apply? Do we need to escape commas or whatever we configured the delimiter to be?

File: src/foam/core/CSV.js [hunk 5]
```diff
@@ -0,0 +1,382 @@
...
+
+    // End a nested object, 1 fewer `__` within object name
+    function end() {
+      this.nestedObjectNames.pop();
+      return this;
+    },
+
+    // TODO: Copied from foam.core.json (move to common library)
+    function nl() {
+      if ( this.nlStr && this.nlStr.length ) {
```
Comment (kgrgreer) [inline src/foam/core/CSV.js]: With JSON, nl's are optional, but with CSV they're mandatory.

File: src/foam/core/CSV.js (hunk 2 above)
Comment (TanayParikh) [reply src/foam/core/CSV.js]: Removed.

File: src/foam/core/CSV.js (hunk 3 above)
Comment (TanayParikh) [reply src/foam/core/CSV.js]: Removed.

File: src/foam/core/CSV.js (hunk 5 above)
Comment (TanayParikh) [reply src/foam/core/CSV.js]: Removed.

File: src/foam/core/CSV.js (hunk 4 above)
Comment (TanayParikh) [reply src/foam/core/CSV.js]: Partially. I removed everything except the escaping of `"`. We still need that as all CSV properties are surrounded by `"` by default (configurable), and it would cause an issue if they are present in the string values.

File: src/foam/core/CSV.js [hunk 6]
```diff
@@ -0,0 +1,359 @@
+/**
+ * @license
+ * Copyright 2017 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+foam.CLASS({
+  package: 'foam.csv',
+  name: 'Outputer',
```
Comment (kgrgreer) [inline src/foam/core/CSV.js]: Correct spelling is Outputter. Two t's.

Comment (TanayParikh) [reply src/foam/core/CSV.js]: Should I correct it in JSON.js as well?

Comment (TanayParikh) [reply src/foam/core/CSV.js]: Fixed for `CSV`

File: src/foam/core/CSV.js [hunk 7]
```diff
@@ -0,0 +1,359 @@
...
+
+    function externalProperty(o, p) {
+      if ( ! this.propertyPredicate(o, p) ) return false;
+      if ( p.isDefaultValue(o[p.name]) ) return false;
+
+      return true;
+    },
+
+    function outputHeaderTitle(o) {
+      this.out(this.includeQuotes ? '"' : '').writePrefix().out(o, this.includeQuotes ? '"' : '', this.delimiter);
```
Comment (kgrgreer) [inline src/foam/core/CSV.js]: If enabled, you'll need to escape " characters.

File: src/foam/core/CSV.js [hunk 8]
```diff
@@ -0,0 +1,359 @@
...
+
+  properties: [
+    {
+      class: 'String',
+      name: 'buf_',
+      value: ''
+    },
+    {
+      class: 'Boolean',
+      name: 'includeQuotes',
```
Comment (kgrgreer) [inline src/foam/core/CSV.js]: I don't think we need this.

File: src/foam/core/CSV.js [hunk 9]
```diff
@@ -0,0 +1,359 @@
...
+      value: true
+    },
+    {
+      class: 'String',
+      name: 'delimiter',
+      value: ','
+    },
+    {
+      class: 'String',
+      name: 'nestedObjectSeperator',
```
Comment (kgrgreer) [inline src/foam/core/CSV.js]: What is this needed for?

File: src/foam/core/CSV.js [hunk 10]
```diff
@@ -0,0 +1,359 @@
...
+      value: '__'
+    },
+    {
+      class: 'StringArray',
+      name: 'nestedObjectNames'
```
Comment (kgrgreer) [inline src/foam/core/CSV.js]: Why is this needed?

File: src/foam/core/CSV.js [hunk 11]
```diff
@@ -0,0 +1,359 @@
...
+    function outputHeader(o) {
+      if ( ! this.outputHeaderRow ) return;
+
+      this.outputPropertyFilteredName(o);
+      this.removeTrailingComma().out(this.nlStr);
+    },
+
+    function removeTrailingComma() {
+      // Removes trailing comma
+      if ( this.buf_.charAt(this.buf_.length - 1) == ',' ) this.buf_ = this.buf_.slice(0, -1);
```
Comment (kgrgreer) [inline src/foam/core/CSV.js]: This is very expensive. Instead, just output a , before each column except the first.

File: src/foam/core/CSV.js [hunk 12]
```diff
@@ -0,0 +1,359 @@
...
+          // Gets and recurses through object properties
+          var ps = o.cls_.getAxiomsByClass(foam.core.Property);
+
+          for ( var i = 0 ; i < ps.length ; i++ ) {
+            this.outputPropertyName(o, ps[i]);
+          }
+        },
+        Array: function(o) { /* Ignore arrays in CSV */ },
+        Function: function(o) { /* Ignore functions in CSV */ },
+        Object: function(o) {
```
Comment (kgrgreer) [inline src/foam/core/CSV.js]: I don't think CSV can support generic Objects since you don't know how many keys they're going to have so then you don't know which column means what.

File: src/foam/core/CSV.js [hunk 13]
```diff
@@ -0,0 +1,359 @@
...
+    },
+
+    function outputHeaderTitle(o) {
+      this.out(this.includeQuotes ? '"' : '').writePrefix().out(o, this.includeQuotes ? '"' : '', this.delimiter);
+    },
+
+    {
+      name: 'outputPropertyFilteredName',
+      code: foam.mmethod({
+        // Is there a `switch` like fall through so the function call isn't unecessarily repeated?
```
Comment (kgrgreer) [inline src/foam/core/CSV.js]: function mmethod(map, opt_defaultMethod) {

File: src/foam/core/CSV.js [hunk 14]
```diff
@@ -0,0 +1,359 @@
...
+      })
+    },
+
+    function outputPropertyName(o, p) {
+      if ( ! this.externalProperty(o, p) ) return;
+
+      // Checks if property is enum, or object with properties of its own
+      // TODO: Is this correct way to check if object has relevant sub-properties
+      if ( foam.core.AbstractEnum.isInstance( o[p.name] ) ) {
+        this.outputPropertyFilteredName(p.name + '__ordinal');
```
Comment (kgrgreer) [inline src/foam/core/CSV.js]: Why is __ordinal required?

File: src/foam/core/CSV.js [hunk 15]
```diff
@@ -0,0 +1,359 @@
...
+      this.removeTrailingComma().out(this.nlStr);
+      ret += this.buf_;
+      this.reset();
+
+      return ret;
+    },
+
+    function externalProperty(o, p) {
+      if ( ! this.propertyPredicate(o, p) ) return false;
+      if ( p.isDefaultValue(o[p.name]) ) return false;
```
Comment (kgrgreer) [inline src/foam/core/CSV.js]: I think we need to output default values. We don't need to do that for JSON because there we're usually loading the data into another FOAM system which already knows the default values. For CSV though, FOAM is very unlikely to be the recipient. Most like it will be a spreadsheet, which doesn't know the default values.

File: src/foam/core/CSV.js (hunk 8 above)
Comment (TanayParikh) [reply src/foam/core/CSV.js]: The configurability or the functionality itself? What should the default be?

File: src/foam/core/CSV.js (hunk 10 above)
Comment (TanayParikh) [reply src/foam/core/CSV.js]: Its a stack of object names. Helps print the header line for nested objects.

File: src/foam/core/CSV.js (hunk 12 above)
Comment (TanayParikh) [reply src/foam/core/CSV.js]: Perfect, I'll remove it. I was unsure on how to test that as well.

File: src/foam/core/CSV.js [hunk 16]
```diff
@@ -0,0 +1,359 @@
...
+  name: 'foam.csv',
+
+  constants: {
+    Compact: foam.csv.Outputter.create({
+      includeQuotes: true
+    }),
+  },
+
+  methods: [
+    function stringify(o) {
```
Comment (kgrgreer) [inline src/foam/core/CSV.js]: I think toCSV() / fromCSV() would be easier to understand.

Comment (TanayParikh) [reply src/foam/core/CSV.js]: Updated, thanks.

File: src/foam/core/CSV.js (hunk 12 above)
Comment (TanayParikh) [reply src/foam/core/CSV.js]: Removed.

File: src/foam/core/CSV.js (hunk 14 above)
Comment (TanayParikh) [reply src/foam/core/CSV.js]: It helps permit nesting logic for the reverse `fromCSV` functionality.

File: src/foam/core/CSV.js (hunk 15 above)
Comment (TanayParikh) [reply src/foam/core/CSV.js]: Updated to output default properties.

File: src/foam/core/CSV.js (hunk 13 above)
Comment (TanayParikh) [reply src/foam/core/CSV.js]: Just tried that, functions inside the map (ex. FObject) seem to be losing scope. Tried binding `this` but still throwing an error. Is there another way to pass in `this` into an `mmethod`?

File: src/foam/core/CSV.js (hunk 14 above)
Comment (kgrgreer) [reply src/foam/core/CSV.js]: How?

File: src/foam/core/CSV.js (hunk 13 above)
Comment (kgrgreer) [reply src/foam/core/CSV.js]: mmethod should preserve 'this' scope

File: src/foam/core/CSV.js (hunk 8 above)
Comment (kgrgreer) [reply src/foam/core/CSV.js]: The functionality.

File: src/foam/core/CSV.js (hunk 10 above)
Comment (kgrgreer) [reply src/foam/core/CSV.js]: But why do you need to store it? This will cause conflicts if you reuse the CSV outputter to output more than one class. So if someone used foam.csv.Compact, which is persistent, then the headers would be corrupted.

File: src/foam/core/CSV.js (hunk 13 above)
Comment (TanayParikh) [reply src/foam/core/CSV.js]: Thanks, updated.

For reference `function (o) { this.xyzMethod(o); }` will need to be used. Arrow functions do not work in this case due to scoping issues, (thanks @kgrgreer).

File: src/foam/core/CSV.js (hunk 11 above)
Comment (TanayParikh) [reply src/foam/core/CSV.js]: Removed.

File: src/foam/core/CSV.js (hunk 8 above)
Comment (TanayParikh) [reply src/foam/core/CSV.js]: Default to include quotes? Or nothing at all?

Comment (kgrgreer) [reply src/foam/core/CSV.js]: I would say nothing at all, but test with Google sheets and/or Excel if you have it to see how it handles importing the two styles.

File: src/foam/core/CSV.js (hunk 14 above)
Comment (TanayParikh) [reply src/foam/core/CSV.js]: Refactored and removed. 	64bce2b

File: src/foam/core/CSV.js (hunk 8 above)
Comment (TanayParikh) [reply src/foam/core/CSV.js]: Google Sheets supports both. Should FOAM do the same?

File: src/foam/core/CSV.js (hunk 10 above)
Comment (TanayParikh) [reply src/foam/core/CSV.js]: Updating.

Is there a way to pass in a param into a `mmethod` (I'll need to pass in the `nestedObjectNames` stack so that it is preserved through recursive calls.

File: src/foam/core/CSV.js (hunk 8 above)
Comment (kgrgreer) [reply src/foam/core/CSV.js]: Just support without quotes. We can add if we ever need it.

File: src/foam/core/CSV.js [hunk 17]
```diff
@@ -0,0 +1,279 @@
...
+      return model;
+    }
+  ]
+});
+
+foam.LIB({
+  name: 'foam.csv',
+
+  constants: {
+    Compact: foam.csv.Outputter.create({
```
Comment (kgrgreer) [inline src/foam/core/CSV.js]: s/Compact/Standard/

Comment (TanayParikh) [reply src/foam/core/CSV.js]: Updated.

File: src/foam/core/CSV.js (hunk 8 above)
Comment (TanayParikh) [reply src/foam/core/CSV.js]: Removed quoted field/value support.

**Comment by TanayParikh** [comment]
@kgrgreer All requested changes have been made, thanks for your help.

File: src/foam/core/CSV.js [hunk 18]
```diff
@@ -0,0 +1,233 @@
...
+
+      this.outputPropertyFilteredName(o, '');
+      this.out(this.nlStr);
+    },
+
+    function toCSV(o) {
+      // Outputs object headers
+      this.outputHeader(o);
+      var ret = this.buf_;
+      this.reset(); // reset to avoid retaining garbage
```
Comment (kgrgreer) [inline src/foam/core/CSV.js]: Calling reset() here loses the contents of the buffer, forcing you to concatenate below, which doubles the amount of garbage that you generate. Wouldn't just reset()-ing at the top of the method be simpler and more efficient?

File: src/foam/core/CSV.js [hunk 19]
```diff
@@ -0,0 +1,233 @@
...
+    function outputHeaderTitle(o, prefix) {
+      this.out(this.buf_.length ? this.delimiter : '')
+          .out(prefix, o);
+    },
+
+    function outputPropertyName(o, p, prefix) {
+      if ( ! this.propertyPredicate(o, p) ) return;
+
+      // Checks if property is enum, or object with properties of its own
+      if ( p.of && ( ! foam.core.AbstractEnum.isInstance( o[p.name] ) ) ) {
```
Comment (kgrgreer) [inline src/foam/core/CSV.js]: p.of will also pickup DAOProperties
Why do you care if it's an Enum? Enum's still only have one value.

File: src/foam/core/CSV.js [hunk 20]
```diff
@@ -0,0 +1,233 @@
...
+        // Appends object name to prefix for CSV Header
+        prefix += p.name + this.nestedObjectSeperator;
+        this.outputPropertyFilteredName(o[p.name], prefix);
+      } else {
+        this.outputPropertyFilteredName(p.name, prefix);
+      }
+    },
+
+    {
+      name: 'outputPropertyFilteredName',
```
Comment (kgrgreer) [inline src/foam/core/CSV.js]: Can you come up with a better name for this method?

File: src/foam/core/CSV.js [hunk 21]
```diff
@@ -0,0 +1,233 @@
...
+      this.output(o[p.name]);
+    },
+
+    function reset() {
+      this.buf_ = '';
+      return this;
+    },
+
+    function outputPrimitive(val) {
+      this.out(this.buf_.length ? this.delimiter : '', val);
```
Comment (kgrgreer) [inline src/foam/core/CSV.js]: I think it would be easier to handle the delimiter where you're doing the looping.

File: src/foam/core/CSV.js [hunk 22]
```diff
@@ -0,0 +1,233 @@
...
+        Null:      function(o) { this.outputPrimitive(null); },
+        String:    function(o) { this.outputPrimitive(o); },
+        FObject:   function(o) {
+          if ( o.outputCSV ) {
+            o.outputCSV(this)
+            return;
+          }
+          
+          // Outputs only the ordinal of ENUM
+          if ( foam.core.AbstractEnum.isInstance(o) ) {
```
Comment (kgrgreer) [inline src/foam/core/CSV.js]: Why are enums treated differently?

Comment (kgrgreer) [reply src/foam/core/CSV.js]: You can put AbstractEnum as one of your mmethod cases.

File: src/foam/core/CSV.js [hunk 23]
```diff
@@ -0,0 +1,233 @@
...
+        Function: function(o) { /* Ignore functions in CSV */ },
+        Object: function(o) { /* Ignore generic objects in CSV */ }
+      }, function(o) { this.outputPrimitive(o); })
+    },
+
+    function fromCSV(className, s) {
+      if ( ! s ) throw 'Invalid CSV input to convert. Arguments must be (className, csvString).'
+      var lines = s.split('\n');
+
+      if ( lines.length == 0 ) throw 'Insufficient CSV Input'
```
Comment (kgrgreer) [inline src/foam/core/CSV.js]: Missing ;

File: src/foam/core/CSV.js [hunk 24]
```diff
@@ -0,0 +1,233 @@
...
+
+      // Trims quotes and splits CSV row into array
+      var props = lines[0].split(',');
+
+      // Account for array of values (TODO?)
+      var values = lines[1].split(',');
```
Comment (kgrgreer) [inline src/foam/core/CSV.js]: What about strings with ,'s in them?

File: src/foam/core/CSV.js [hunk 25]
```diff
@@ -0,0 +1,233 @@
...
+  constants: {
+    Standard: foam.csv.Outputter.create(),
+  },
+
+  methods: [
+    function toCSV(o) {
+      return foam.csv.Standard.toCSV(o);
+    },
+
+    function fromCSV(className, s) {
```
Comment (kgrgreer) [inline src/foam/core/CSV.js]: Why pass the className rather than just the class/model itself?

File: src/foam/core/CSV.js (hunk 24 above)
Comment (kgrgreer) [inline src/foam/core/CSV.js]: Shouldn't it be valid to have no lines?

Comment (kgrgreer) [inline src/foam/core/CSV.js]: Where/how do you process the lines after the first line?

File: src/foam/core/CSV.js [hunk 26]
```diff
@@ -0,0 +1,233 @@
...
+                                          values.slice(i, j), prop.of.id));
+              
+              i = j - 1;
+              break;
+            }
+          }
+        // Adds regular prop
+        } else {
+          var prop = model.cls_.getAxiomByName(p);
+          prop.set(model, prop.of ? foam.lookup(prop.of.id).create({ ordinal: v }) : v);
```
Comment (kgrgreer) [inline src/foam/core/CSV.js]: Isn't foam.lookup(prop.of.id) the same as just prop.of?

File: src/foam/core/CSV.js (hunk 24 above)
Comment (TanayParikh) [reply src/foam/core/CSV.js]: There isn't really a way to handle that unless we use a special delimiter (~), or we surround the fields with a symbol ("), as far as I know. What would you recommend?

Comment (TanayParikh) [reply src/foam/core/CSV.js]: What would the output be in that case? Empty FObject?

Comment (TanayParikh) [reply src/foam/core/CSV.js]: I process only the first (header) line and the second (value) line. I was not sure if we were to add support for multiple value lines, so I left it as a potential TODO on line 170. I can add it in if required.

File: src/foam/core/CSV.js (hunk 23 above)
Comment (TanayParikh) [reply src/foam/core/CSV.js]: Fixed.

File: src/foam/core/CSV.js (hunk 24 above)
Comment (kgrgreer) [reply src/foam/core/CSV.js]: I don't think there should be any return value. You need to return more than one value, one per row, so I would just have this method return void and pass in a Sink as a third argument for it to put() objects to.

Comment (kgrgreer) [reply src/foam/core/CSV.js]: Yes, that's absolutely required. CSV files usually have many lines.

File: src/foam/core/CSV.js (hunk 21 above)
Comment (TanayParikh) [reply src/foam/core/CSV.js]: Added `first` flag, as done in `JSON.js`.

File: src/foam/core/CSV.js (hunk 18 above)
Comment (TanayParikh) [reply src/foam/core/CSV.js]: Fixed.

File: src/foam/core/CSV.js (hunk 22 above)
Comment (TanayParikh) [reply src/foam/core/CSV.js]: Added to mmethod case.

File: src/foam/core/CSV.js (hunk 26 above)
Comment (TanayParikh) [reply src/foam/core/CSV.js]: Updated, thanks.

File: src/foam/core/CSV.js (hunk 24 above)
Comment (kgrgreer) [reply src/foam/core/CSV.js]: Google Sheets wraps the whole value in ""'s.

Comment (TanayParikh) [reply src/foam/core/CSV.js]: Implemented Sink.

Comment (TanayParikh) [reply src/foam/core/CSV.js]: Multiple lines are processed, and put into Sink as recommended.

File: src/foam/core/CSV.js (hunk 19 above)
Comment (TanayParikh) [reply src/foam/core/CSV.js]: `p.of` Was used to check if the property was an 'inner object' (with properties of its own). Is there a better/more conventional way to achieve this?

I'm handling Enum's slightly differently to ensure only the ordinal is printed. Otherwise, if treated like a regular object, the ordinal, label and name are all printed. (`JSON.js` prints only the ordinal).

File: src/foam/core/CSV.js [hunk 27]
```diff
@@ -166,11 +166,14 @@ foam.CLASS({
       // Trims quotes and splits CSV row into array
       var props = lines[0].split(',');
 
-      // Account for array of values (TODO?)
-      var values = lines[1].split(',');
+      for (var i = 1; i < lines.length; ++i) {
```
Comment (kgrgreer) [inline src/foam/core/CSV.js]: spaces before ;'s

Comment (TanayParikh) [reply src/foam/core/CSV.js]: Fixed, thanks.

File: src/foam/core/CSV.js (hunk 20 above)
Comment (TanayParikh) [reply src/foam/core/CSV.js]: Renamed to `outputPropertyName`.

**Comment by TanayParikh** [comment]
Please note, the sink is currently being converted to an array, and then returned. This functionality will be removed on the final build once testing is completed. 

**Edit:** Removed 34aa8c2. Kept return of sink (could be useful in certain cases).

File: src/foam/core/CSV.js (hunk 24 above)
Comment (TanayParikh) [reply src/foam/core/CSV.js]: Added custom CFG parser.

FObject -> CSV
- Fields with commas are surrounded in quotes. Quotes within the string itself are escaped with a second quote char. (Google Sheets approach)

File: src/foam/core/CSV.js (hunk 25 above)
Comment (TanayParikh) [reply src/foam/core/CSV.js]: I tried passing in the model (via `foam.lookup('fullQualifier')`), however, it gives the following error: `customModel.create is not a function`

When I pass in just the className, `foam.lookup('fullQualifier')` is exactly how I convert it into a model. I’m not sure what’s causing the error in that case.

File: src/foam/lib/csv/CSV.js [hunk 28]
```diff
@@ -0,0 +1,256 @@
...
+    function outputHeaderTitle(o, prefix, first) {
+      this.out(first ? '' : this.delimiter)
+          .out(this.escapeString(prefix + o));
+    },
+
+    function outputPropertyName_(o, p, prefix, first) {
+      if ( ! this.propertyPredicate(o, p) ) return;
+
+      // Checks if property is enum, or object with properties of its own
+      if ( p.of && ( ! foam.core.AbstractEnum.isInstance( o[p.name] ) ) ) {
```
Comment (kgrgreer) [inline src/foam/lib/csv/CSV.js]: The above comment and the code don't agree.

Comment (kgrgreer) [reply src/foam/lib/csv/CSV.js]: Better to just check if the property is an FObjectProperty than looking for 'of'.

File: src/foam/lib/csv/CSV.js [hunk 29]
```diff
@@ -0,0 +1,256 @@
...
+          }
+        },
+        Array: function(o) { /* Ignore arrays in CSV */ },
+        Function: function(o) { /* Ignore functions in CSV */ },
+        Object: function(o) { /* Ignore generic objects in CSV */ }
+      }, function(o, prefix, first) { this.outputHeaderTitle(o, prefix, first); })
+    },
+
+    function outputProperty(o, p, first) {
+      if ( ! this.propertyPredicate(o, p) ) return;
```
Comment (kgrgreer) [inline src/foam/lib/csv/CSV.js]: A little more straight-forward, and less code, to just do:
if ( this.propertyPredicate(o, p) ) this.output(o[p.name], first);

File: src/foam/lib/csv/CSV.js [hunk 30]
```diff
@@ -0,0 +1,256 @@
...
+      }
+      
+      return s;
+    },
+
+    {
+      name: 'output',
+      code: foam.mmethod({
+        Undefined:    function(o, first) { this.outputPrimitive(this.undefinedStr, first); },
+        Null:         function(o, first) { this.outputPrimitive(null, first); },
```
Comment (kgrgreer) [inline src/foam/lib/csv/CSV.js]: Is this line redundant?

File: src/foam/lib/csv/CSV.js [hunk 31]
```diff
@@ -0,0 +1,256 @@
...
+        sink.put(this.createModel(props, values, className));
+      }
+
+      return sink;
+    },
+
+    function splitIntoValues(csvString) {
+      if ( ( csvString == undefined ) || ( csvString.length == 0) ) return [];
+
+      var parser = foam.lookup('foam.lib.csv.CSVParser').create();
```
Comment (kgrgreer) [inline src/foam/lib/csv/CSV.js]: Why do you have to look this up?

File: src/foam/lib/csv/CSV.js [hunk 32]
```diff
@@ -0,0 +1,256 @@
...
+            this.outputProperty(o, ps[i], (i == 0 && first));
+          }
+        },
+        Array:        function(o) { /* Ignore arrays in CSV */ },
+        Function:     function(o) { /* Ignore functions in CSV */ },
+        Object:       function(o) { /* Ignore generic objects in CSV */ }
+      }, function(o, first) { this.outputPrimitive(o, first); })
+    },
+
+    function fromCSV(className, s, sink) {
```
Comment (kgrgreer) [inline src/foam/lib/csv/CSV.js]: Why passing className rather than just the class?

File: src/foam/lib/csv/CSV.js [hunk 33]
```diff
@@ -0,0 +1,256 @@
...
+    },
+
+    function splitIntoValues(csvString) {
+      if ( ( csvString == undefined ) || ( csvString.length == 0) ) return [];
+
+      var parser = foam.lookup('foam.lib.csv.CSVParser').create();
+      return parser.parseString(csvString).map(field => field.value == undefined ? '' : field.value);
+    },
+
+    function createModel(props, values, className) {
```
Comment (kgrgreer) [inline src/foam/lib/csv/CSV.js]: Why passing className instead of just class?

File: src/foam/lib/csv/CSV.js [hunk 34]
```diff
@@ -0,0 +1,256 @@
...
+      foam.assert(props.length == values.length,
+        'Invalid CSV Input, header and value rows must be the same number of cells');
+
+      var model = foam.lookup(className).create();
+
+      for ( var i = 0 ; i < props.length ; ++i ) {
```
Comment (kgrgreer) [inline src/foam/lib/csv/CSV.js]: i++

File: src/foam/lib/csv/CSV.js [hunk 35]
```diff
@@ -0,0 +1,256 @@
...
+        var p = props[i];
+        var v = values[i];
+
+        // Adds nested prop
+        if ( p.includes('__') ) {
```
Comment (kgrgreer) [inline src/foam/lib/csv/CSV.js]: Shouldn't this use nestedObjectSeperator instead?

File: src/foam/lib/csv/CSV.js [hunk 36]
```diff
@@ -0,0 +1,256 @@
...
+          p = p.split('__');
```
Comment (kgrgreer) [inline src/foam/lib/csv/CSV.js]: Shouldn't this use nestedObjectSeperator instead?

File: src/foam/lib/csv/CSV.js [hunk 37]
```diff
@@ -0,0 +1,256 @@
...
+          foam.assert(p.length >= 2, 'Invalid CSV object nesting, properties of inner objects are identified by `__`');
+          var prefix = p[0] + '__';
```
Comment (kgrgreer) [inline src/foam/lib/csv/CSV.js]: Shouldn't this use nestedObjectSeperator instead?

File: src/foam/lib/csv/CSV.js (hunk 33 above)
Comment (TanayParikh) [reply src/foam/lib/csv/CSV.js]: I tried passing in the model (via `foam.lookup('fullQualifier')`), however, it gives the following error: `customModel.create is not a function`

When I pass in just the className, `foam.lookup('fullQualifier')` is exactly how I convert it into a model. I’m not sure what’s causing the error in that case.

File: src/foam/lib/csv/CSV.js [hunk 38]
```diff
@@ -0,0 +1,256 @@
...
+  constants: {
+    Standard: foam.lib.csv.Outputter.create(),
+  },
+
+  methods: [
+    function toCSV(o) {
+      return foam.lib.csv.Standard.toCSV(o);
+    },
+
+    function fromCSV(className, csvString, sink) {
```
Comment (kgrgreer) [inline src/foam/lib/csv/CSV.js]: Why pass className instead of just class?

File: src/foam/lib/csv/CSV.js (hunk 37 above)
Comment (TanayParikh) [reply src/foam/lib/csv/CSV.js]: Yes, updated.

File: src/foam/lib/csv/CSV.js (hunk 36 above)
Comment (TanayParikh) [reply src/foam/lib/csv/CSV.js]: Yes, updated.

File: src/foam/lib/csv/CSV.js (hunk 35 above)
Comment (TanayParikh) [reply src/foam/lib/csv/CSV.js]: Yes, updated.

File: src/foam/lib/csv/CSV.js (hunk 34 above)
Comment (TanayParikh) [reply src/foam/lib/csv/CSV.js]: Updated.

File: src/foam/lib/csv/CSV.js (hunk 31 above)
Comment (TanayParikh) [reply src/foam/lib/csv/CSV.js]: Was redundant, removed.

File: src/foam/lib/csv/CSV.js (hunk 30 above)
Comment (TanayParikh) [reply src/foam/lib/csv/CSV.js]: Removed.

File: src/foam/lib/csv/CSV.js (hunk 29 above)
Comment (TanayParikh) [reply src/foam/lib/csv/CSV.js]: Thanks, updated.

File: src/foam/lib/csv/CSV.js (hunk 28 above)
Comment (TanayParikh) [reply src/foam/lib/csv/CSV.js]: Reworded to better reflect the code.

File: src/foam/lib/csv/CSV.js (hunk 33 above)
Comment (kgrgreer) [reply src/foam/lib/csv/CSV.js]: When you pass in the class (not the model, which you actually rarely use in FOAM2) you should be able to call create on it, so let's switch to that and then diagnose your "is not a function" error. Also, why do you need to lookup? Just declare that you require: the class then refer to it as this.Class. This has the advantage that it lets the dynamic code-loading support know what classes it should load when load the calling class.

Comment (TanayParikh) [reply src/foam/lib/csv/CSV.js]: Thanks, that really helped! I've updated accordingly.

File: src/foam/lib/csv/CSV.js [hunk 39]
```diff
@@ -0,0 +1,257 @@
...
+    },
+
+    function outputPropertyName_(o, p, prefix, first) {
+      if ( ! this.propertyPredicate(o, p) ) return;
+
+      // Checks if property is an inner object (and not an enum)
+      if ( p.of && ( ! foam.core.AbstractEnum.isInstance( o[p.name] ) ) ) {
+        // Gets new empty object if FObjectProperty is currently undefined
+        // Done to permit appropriate headers for multi-line CSVs (multiple objects to convert)
+        if ( o[p.name] == undefined ) o[p.name] = foam.lookup(p.of.id);
```
Comment (kgrgreer) [inline src/foam/lib/csv/CSV.js]: foam.lookup(p.of.id) is the same as just p.of, isn't it?

File: src/foam/lib/csv/CSV.js [hunk 40]
```diff
@@ -0,0 +1,257 @@
...
+    function outputHeaderTitle(o, prefix, first) {
+      this.out(first ? '' : this.delimiter)
+          .out(this.escapeString(prefix + o));
+    },
+
+    function outputPropertyName_(o, p, prefix, first) {
+      if ( ! this.propertyPredicate(o, p) ) return;
+
+      // Checks if property is an inner object (and not an enum)
+      if ( p.of && ( ! foam.core.AbstractEnum.isInstance( o[p.name] ) ) ) {
```
Comment (kgrgreer) [inline src/foam/lib/csv/CSV.js]: Can't you just check if p is an FObjectProperty?
